src/main/java/graphql/execution/instrumentation/dataloader/FieldLevelTrackingApproach.java

```diff
@@ -30,65 +29,67 @@
     private final Supplier<DataLoaderRegistry> dataLoaderRegistrySupplier;
     private final Logger log;
 
-    private static class CallStack implements InstrumentationState {
+    private static class FieldTracking {
 
-        private final Map<Integer, Integer> expectedFetchCountPerLevel = new LinkedHashMap<>();
-        private final Map<Integer, Integer> fetchCountPerLevel = new LinkedHashMap<>();
-        private final Map<Integer, Integer> expectedStrategyCallsPerLevel = new LinkedHashMap<>();
-        private final Map<Integer, Integer> happenedStrategyCallsPerLevel = new LinkedHashMap<>();
-        private final Map<Integer, Integer> happenedOnFieldValueCallsPerLevel = new LinkedHashMap<>();
+        final AtomicInteger expectedFetchCountPerLevel = new AtomicInteger();
+        final AtomicInteger fetchCountPerLevel = new AtomicInteger();
+        final AtomicInteger expectedStrategyCallsPerLevel = new AtomicInteger();
+        final AtomicInteger happenedStrategyCallsPerLevel = new AtomicInteger();
+        final AtomicInteger happenedOnFieldValueCallsPerLevel = new AtomicInteger();
+    }
```

Comment:
One challenge with this change is that all the CallStack state variables are now independent.  Before they could only be changed in unison - so if `expectedStrategyCallsPerLevel` some how depends on the value of `happenedStrategyCallsPerLevel` for a consistent view then the synchronized block gave you that

Now they thread safe atomic values that can be changed independently of each other.

My problem is - I am not sure IF there is in fact a dependency problem of not.  So at this stage is theoretical to say this is problematic.

We need to try and start at the code and see if there is any dependent state between the variables

Reply:
Yeah for example we had this

```
        synchronized (callStack) {
            callStack.increaseExpectedFetchCount(curLevel, fieldCount);
            callStack.increaseHappenedStrategyCalls(curLevel);
        }
```
This nominally sets two values atomically - they will always be in sync with each other at this stage.  That said its in `beginExecutionStrategy` so its definitely not in multi threaded stages yet (on the data fetching stage is possibly multithreaded

Reply:
Another case

```
//
    // thread safety : called with synchronised(callStack)
    //
    private boolean handleOnFieldValuesInfo(List<FieldValueInfo> fieldValueInfoList, CallStack callStack, int curLevel) {
        callStack.increaseHappenedOnFieldValueCalls(curLevel);
        int expectedStrategyCalls = 0;
        for (FieldValueInfo fieldValueInfo : fieldValueInfoList) {
            if (fieldValueInfo.getCompleteValueType() == FieldValueInfo.CompleteValueType.OBJECT) {
                expectedStrategyCalls++;
            } else if (fieldValueInfo.getCompleteValueType() == FieldValueInfo.CompleteValueType.LIST) {
                expectedStrategyCalls += getCountForList(fieldValueInfo);
            }
        }
        callStack.increaseExpectedStrategyCalls(curLevel + 1, expectedStrategyCalls);
        return dispatchIfNeeded(callStack, curLevel + 1);
    }
    ```
    
we have multiple calls into the callstack that is both incrementing and comparing the state values.  I am concerned we will get subtle bugs because the values are not consistent.  

And with DataLoader the costs are high - IF we miss a dispatch that should have happened and no other fields cause this to reevaluate - then we get a fetch deadlock

Reply:
oh I see, and this kind of bugs are quite hard to track (usually only in production!).

Reply:
Yes very hard to re-produce and cause - its ONLY with concurrent access (withn the same request) that can make the numbers fluctuate